### PR TITLE
Updated epaddr to use references with lifetimes instead of Rc.

### PR DESCRIPTION
**Pull request by hestela**
Using macros in defined in other files/mods/crates is a bit weird in rust atm.
The macro i added is not often, it was more of a test to see if i could namespace macros in thier own file.

@acmcarther

### CODE REVIEW

File: ep/src/lib.rs [hunk 1]
```diff
@@ -47,21 +50,20 @@ pub struct Epaddr {
...
+  mctxt_prev: Option<&'a Epaddr<'a>>,
+  mctxt_next: Option<&'a Epaddr<'a>>
 }
 
 
 impl Ep {
   // TODO: change job_key to rust uuid, add psm_ep
-  pub fn open(job_key: u64, ep_opts: EpOpts) -> Result<Box<(Ep, Epaddr)>, Error> {
+  pub fn open<'a>(job_key: u64, ep_opts: EpOpts) -> Result<(Ep, &'a Epaddr<'a>), Error> {
     Err(Error { error: Error_type::PSM_ERROR_LAST, error_str: "send help"})
```
Comment (acmcarther) [inline ep/src/lib.rs:70]: Can we get `new` defined on `Error`?

``` rust
     Err(Error::new(Error_type::PSM_ERROR_LAST, "send help"}))
```

File: ep/src/lib.rs [hunk 2]
```diff
@@ -69,13 +71,13 @@ impl Ep {
     Ok(())
   }
 
-  pub fn connect(ep: Ep, epids: &Vec<Epid>, epid_masks: &Vec<isize>, timeout: i64) -> Result<Box<Vec<Epaddr>>, Box<Vec<Error>>> {
+  pub fn connect<'a>(ep: Ep, epids: Box<Vec<Epid>>, epid_masks: &Vec<isize>, timeout: i64) -> Result<Box<Vec<&'a Epaddr<'a>>>, Box<Vec<Error>>> {
     let dummy: Vec<Error> = Vec::new();
     Err(Box::new(dummy))
```
Comment (acmcarther) [inline ep/src/lib.rs]: You don't need the types for dummy, or really dummy at all I think

``` rust
Err(Box::new(vec!))
```

File: ep/src/lib.rs [hunk 3]
```diff
@@ -47,21 +50,20 @@ pub struct Epaddr {
   mctxt_nsconn: usize,
   mctxt_send_seqnum: u16,
   mctxt_recv_seqnum: u16,
-  mctxt_current: Option<Rc<Epaddr>>,
+  mctxt_current: Option<&'a Epaddr<'a>>,
```
Comment (acmcarther) [inline ep/src/lib.rs:56]: From my coworker, it sounds like you can use an enum variant for this

http://rustbyexample.com/enum.html has an example with a linked list. It uses a box, but you should be able to use a reference. Might need lifetimes.

The box actually might be the most sane thing.... hmmm.

File: ep/src/lib.rs [hunk 4]
```diff
@@ -69,14 +71,13 @@ impl Ep {
...
-    Err(Box::new(dummy))
+  pub fn connect<'a>(ep: Ep, epids: Box<Vec<Epid>>, epid_masks: &Vec<isize>, timeout: i64) -> Result<Box<Vec<&'a Epaddr<'a>>>, Box<Vec<Error>>> {
+    Err(Box::new(vec!()))
   }
 }
 
-impl Epaddr {
-
+impl <'a>Epaddr <'a> {
+  // TODO: figure out what fields are commonly used when epaddr is constructed
```
Comment (hestela) [inline ep/src/lib.rs:83]: These are often contructed using different amounts of parameters.

Comment (acmcarther) [inline ep/src/lib.rs:83]: Maybe a bunch of specialized constructors with intuitive names?

File: ep/src/lib.rs (hunk 3 above)
Comment (hestela) [inline ep/src/lib.rs:56]: I just noticed that this particular field is sometimes used as a reference to a single epaddr and then other times pointing to the first item in an array on the stack..... gj C. Going to have to investigate more if they treat it that way. Ill put a comment about this.

**Comment by acmcarther** [comment]
Oh man that is gross.

My comment meant put the `-> Error` on the function line.

as in `fn stuff() -> Thing`

Also enum type names probs should be `ThisCase` not snake.

**Comment by hestela** [comment]
was lazy and pushed to master 1c5960a9d4c677ced67eeae2dbba6be051ca8578

**Pull request merged**